domain/secretbackend/service/service.go

```diff
@@ -700,9 +700,53 @@ func (s *Service) RotateBackendToken(ctx context.Context, backendID string) erro
 	return errors.Trace(err)
 }
 
-// SetModelSecretBackend sets the secret backend for the given model UUID.
+// GetModelSecretBackend returns the secret backend name for the given model UUID, returning an error
+// satisfying [modelerrors.NotFound] if the model provided does not exist.
+func (s *Service) GetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID) (string, error) {
+	backendName, modelType, err := s.st.GetModelSecretBackend(ctx, modelUUID)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	switch modelType {
+	case coremodel.IAAS:
+		if backendName == provider.Internal {
+			backendName = provider.Auto
+		}
+	case coremodel.CAAS:
+		if backendName == kubernetes.BackendName {
+			backendName = provider.Auto
+		}
+	default:
+		// Should never happen.
+		return "", errors.NotValidf("model type %q", modelType)
+	}
+	return backendName, nil
+}
+
+// SetModelSecretBackend sets the secret backend config for the given model UUID,
+// returning an error satisfying [secretbackenderrors.NotFound] if the backend provided does not exist,
+// returning an error satisfying [modelerrors.NotFound] if the model provided does not exist,
+// returning an error satisfying [secretbackenderrors.NotValid] if the backend name provided is not valid.
 func (s *Service) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error {
-	return s.st.SetModelSecretBackend(ctx, modelUUID, backendName)
+	if backendName == "" {
+		return fmt.Errorf("missing backend name%w", errors.Hide(secretbackenderrors.NotValid))
+	}
+	if backendName == provider.Internal || backendName == kubernetes.BackendName {
+		return fmt.Errorf("secret backend name %q not valid%w", backendName, errors.Hide(secretbackenderrors.NotValid))
+	}
+	return s.st.SetModelSecretBackend(ctx, modelUUID, func(modelType coremodel.ModelType) (string, error) {
```

Comment:
I don't agree with this approach. It makes no sense why the state layer has to call back up in a closure to make decisions on what the backend should or shouldn't be. I even believe this is a change from what I fixed up several weeks ago when this logic used to live in the state layer.

Model type doesn't change over the life of a model. You should do all of this work in the service then just call state with the value that needs to be set.